File: modules/domain/src/main/java/com/streamfirst/iceberg/hybrid/domain/StorageLocation.java
package com.streamfirst.iceberg.hybrid.domain;

import java.util.Objects;

/**
 * Represents a storage endpoint in a specific region.
 * Each region can have multiple storage locations for different purposes
 * (primary data, backup, temporary files, etc.).
 * 
 * @param region the geographic region where this storage is located
 * @param uri the storage URI (e.g., "s3://bucket-name", "hdfs://cluster/path")
 * @param type the storage type/purpose (e.g., "primary", "backup", "temp")
 */
public record StorageLocation(Region region, String uri, String type) {
    public StorageLocation {
        Objects.requireNonNull(region, "Region cannot be null");
        Objects.requireNonNull(uri, "URI cannot be null");
        Objects.requireNonNull(type, "Type cannot be null");
        
        if (uri.trim().isEmpty()) {
            throw new IllegalArgumentException("URI cannot be empty");
        }
    }
}

File: modules/domain/src/main/scala/com/streamfirst/iceberg/hybrid/domain/StorageLocation.scala
package com.streamfirst.iceberg.hybrid.domain

/** Represents a storage location in a specific region. Contains all information needed to access
  * the storage system in that region.
  */
final case class StorageLocation(
  region: Region,
  endpoint: String,
  bucketName: String,
  pathPrefix: Option[String] = None
):
  def basePath: StoragePath =
    val prefix = pathPrefix.map(_ + "/").getOrElse("")
    StoragePath(s"s3://$bucketName/$prefix")

  def pathFor(relativePath: String): StoragePath =
    basePath / relativePath

object StorageLocation:
  def s3(region: Region, bucket: String, prefix: Option[String] = None): StorageLocation =
    val endpoint = s"s3.${region.id}.amazonaws.com"
    StorageLocation(region, endpoint, bucket, prefix)
